### ISSUE TITLE
[EPCO-20] Training process will result in a smoother final distance field

### ISSUE THREAD

**Issue by hatfield-c** [hatfield-c/epco#20]
Presently the training process results in a distance field which is noisy and choppy. The model may be able to learn the binary separation, but the final distance field is all but impossible to intelligently traverse during active inference.

The best way to address this, without inventing an entirely new means of training beyond gradient descent (or potentially a new loss function), would be to simultaneously train on binary separation as well as train the model to produce distance values such that points closer to the decision boundary are lower than points further from the decision boundary.

**Comment by hatfield-c** [comment]
The binary cross entropy loss function can be repurposed for this task. Let x1 and x2 be two points with positive distance, and y = 1 when we want x1<x2 and y = 0 when we want x2<x1. Then the loss function becomes:
- y * log( Sigmoid ( N(x2) - N(x1) ) ) + (1 - y) * (1 - log( Sigmoid( N(x1) - N(x2) ) ))

A similar loss function will need to be created when training points in negative space.

**Comment by hatfield-c** [comment]
Because the decision boundary is constantly changing with the training process, we will unfortunately not be able to pre-compile the training data for the smoothing process.

The current plan to generate this training data on the fly is such:
- Choose a batch of points from the training data used by the separation training process, and potentially throw in some new random points as well.
- Connect these points into a web, such that all points are connected to the three points closest to them.
- Divide the points into "zones" or "strands", such that a low point to the left side of a hill is only compared to other points on the left side of the hill, and the same with low points on the right side.
    - Each strand should begin with a negative point followed by all positive points
- Starting at the point that comes right after the beginning negative point, walk down the strand two at a time
- While walking, compile:
    - The two points being considered
    - Their ordering (which should be lower vs. higher)
    - Whether they are in the positive or negative distance field
- Pass the compiled data through the modified forward function
- Pass the result to the loss function

**Comment by hatfield-c** [comment]
The idea of trying to train the surface of N to always rise and then fall is proving incredibly troublesome, as the risk of missing small pockets of negative space in a sea of positive space is too great.

An easier means of smoothing N might be instead to train such that there are no local minima inside of positive space and no local maxima inside of negative space.

**Comment by hatfield-c** [comment]
Potential smoothing method via elimination of contextual optima:
- Pick a random query point in positive space
- Pick a random local sample point within a certain distance that are is in positive space.
- Pick a sample point in the *exact* opposite direction from the previous point, which should also be in positive space
- Perform one of the following actions:
    - If the query point is above one sample point while below another, then there is no issue, and this data point should be discarded
    - If the query point is above both points, then there is no issue, and this data point should be discarded
    - If the query point is below both points, then there is a local minima around the query point. and the data point is added to the batch
- When a data point is added to the batch, it is added as so:
    - The query point is put into the first position, which is the "larger" position
    - The The sample point is put into the second position, which is the "smaller" position
- The data points are then fed to the new loss function, which is based on the previous custom BCE loss:
    - log( Sigmoid( N(query) - N(sample) ) )

A similar process can be used for points in negative space.

**Comment by hatfield-c** [comment]
Alternative smoothing method from Dr. Yen:
- Train the noisy N
- Find a degree 2 hyper-polynomial that fits the current shape of N
- Use MSE to have N fit the shape of the polynomial
- Repeat Phase 1 - Phase 2 alternations until satisfied

**Comment by hatfield-c** [comment]
This will be accomplished with the voxel neural network architecture. While a voxel has a lot of limitations on the volumes it can fill, by using multiple voxels we can arbitrarily fit any shape within a desired margin of error.

**Comment by hatfield-c** [comment]
This issue is thus complete and will now be closed.
